<a href="https://colab.research.google.com/github/jumunozle/AlgorithmsUN2024II/blob/main/Lab3/jumunozle_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

2 3
5

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Overwriting output1.txt


In [ ]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
9

Overwriting output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

7 9
16


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

5
1 2 3 4 5
20


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Overwriting input0.txt


In [ ]:
%%writefile output0.txt
90

Overwriting output0.txt


In [ ]:
!./max_pairwise_product_full < input0.txt

0


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

0


In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Writing input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Overwriting output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

0
1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
4
7 12 23 34

Overwriting input4.txt


In [ ]:
%%writefile output4.txt
782

Overwriting output4.txt


In [ ]:
%%writefile input5.txt
5
2 4 6 8 10

Overwriting input5.txt


In [ ]:
%%writefile output5.txt
80

Overwriting output5.txt


In [ ]:
%%writefile input6.txt
6
15 30 45 60 75 90

Overwriting input6.txt


In [ ]:
%%writefile output6.txt
6750

Overwriting output6.txt


In [ ]:
%%writefile input7.txt
7
3 6 9 12 15 18 21

Overwriting input7.txt


In [ ]:
%%writefile output7.txt
378

Overwriting output7.txt


In [ ]:
%%writefile input8.txt
8
11 22 33 44 55 66 77 88

Overwriting input8.txt


In [ ]:
%%writefile output8.txt
6776

Overwriting output8.txt


In [ ]:
%%writefile input9.txt
9
5 10 15 20 25 30 35 40 45

Overwriting input9.txt


In [ ]:
%%writefile output9.txt
1800

Overwriting output9.txt


In [ ]:
%%writefile input10.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input10.txt


In [ ]:
%%writefile output10.txt
90

Overwriting output10.txt


In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {4..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!chmod u+x stress_test.sh
!./stress_test.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Overwriting output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Overwriting output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	5m0.951s
user	4m57.441s
sys	0m0.155s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m2.164s
user	4m58.247s
sys	0m0.107s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt
!./max_pairwise_product_long_full_execints < input2.txt
!./max_pairwise_product_long_full_execints < input3.txt

600
571
231


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
11 + 6*(200000) + 5.5*(200000)**2

220001200011.0

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full < input11.txt

2147483646

real	5m8.157s
user	5m4.783s
sys	0m0.129s


In [ ]:
# Estimated time by instruction

304/200003177092


1.5199758544843627e-09

In [ ]:
# Estimated number of instructions by second

200003177092/304

657905187.8026316

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full < input12.txt

2147483646

real	5m8.892s
user	5m4.806s
sys	0m0.150s


### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m1.322s
user	0m1.263s
sys	0m0.003s


## Python

In [ ]:
%%writefile max_pairwise_product_full.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(0,n):                #n + 2n**2
    if i!=j:                            #n**2
      auxProduct = list[i]*list[j]      #n**2 -n
    if maxProduct < auxProduct:         #n**2 -n
      maxProduct = auxProduct           #[1, n**2/2 - n/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + 3n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound

Overwriting max_pairwise_product_full.py


In [ ]:
!chmod u+x max_pairwise_product_full.py

!python max_pairwise_product_full.py < input3.txt > output3.txt
!python max_pairwise_product_full.py < input4.txt > output4.txt
!python max_pairwise_product_full.py < input5.txt > output5.txt
!python max_pairwise_product_full.py < input6.txt > output6.txt
!python max_pairwise_product_full.py < input7.txt > output7.txt
!python max_pairwise_product_full.py < input8.txt > output8.txt
!python max_pairwise_product_full.py < input9.txt > output9.txt
!python max_pairwise_product_full.py < input10.txt > output10.txt

In [ ]:
fileName = "inputPy.txt"

with open(fileName, "w") as file:
    file.write("20000"+ "\n")
    for i in range(1,20001):
        file.write(str(i) + " ")

In [ ]:
import time

start_time = time.time()
!python max_pairwise_product_full.py < inputPy.txt

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time} segs")
print(f"Execution time: {execution_time/60} min")

399980000
Execution time: 136.16147661209106 segs
Execution time: 2.269357943534851 min


In [ ]:
%%writefile max_pairwise_product_full_count.py
op =0
n = int(input())                        #1
op+=1
list = input().split()                  #1
op+=1
list = [int(x) for x in list]           #n
op+=n


maxProduct = 0                          #1
op+=1
auxProduct = 0                          #1
op+=1
for i in range(0,n):                    #1+2n
  op+=1
  op+=2
  for j in range(0,n):                  #n + 2n**2
    op+=2
    op+=1
    if i!=j:                            #n**2
      auxProduct = list[i]*list[j]      #n**2 -n
      op+=1
      op+=1
    if maxProduct < auxProduct:         #n**2 -n
      maxProduct = auxProduct           #n * (n - 1) / 3
      op+=1

                                        # -------------- for worst case
print("Result: "+str(maxProduct))
print("Number of instructions: "+str(op))

Writing max_pairwise_product_full_count.py


The number of instructions executed by the algooirithm with a list of size $n$
 is
 $$T=5+2n+5(n**2)+(n(n-1)/3)$$
 It depends on the size of the lists and the order of the list.

For the input1.txt

10
1 2 3 4 5 6 7 8 9 10
$$T=5+(20)+5(10)(10)+90/3=555$$

In [ ]:
!python max_pairwise_product_full_count.py < input1.txt

Result: 90
Number of instructions: 554


The number of instructions executed by the algooirithm with a list of size
 It depends on the size of the lists and the order of the list.

For the inputPy.txt
$$T=5+(60000)+3(20000)(20000)+2(20000)(20000)-20000+199990/3=2000106668$$

In [ ]:
!python max_pairwise_product_full_count.py < inputPy.txt

Result: 399980000
Number of instructions: 2000261148


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600
CPU times: user 27.9 ms, sys: 3.94 ms, total: 31.8 ms
Wall time: 1.52 s


In [ ]:
%%time
!python max_pairwise_product_full_count.py < input1.txt

Result: 90
Number of instructions: 554
CPU times: user 7.98 ms, sys: 0 ns, total: 7.98 ms
Wall time: 106 ms


In [ ]:
time_per_ins = 0.0197 / 589
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00646 / 554
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 3.344651952461799e-05
Instructions per second C++: 29898.477157360412
Time per instruction Python: 1.1660649819494584e-05
Instructions per second Python: 85758.51393188855


If we replace T in the previous equations we have:

Max n that c++ program can solve in a second: 74.5

Max n that python program can solve in a second: 126.64

# For the follwing solutions required  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input11.txt

39999800000

real	2m25.623s
user	2m21.670s
sys	0m0.116s


In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {1..10}
do
    ./max_pairwise_product_long_full < input$i.txt > output$i.txt
done

Overwriting stress_test.sh


In [ ]:
print(11 + 7*(200000) + 2*(200000)*(199999) + 1, 11 + 7*(200000) + 2.5*(200000)*(199999))

80001000012 100000900011.0


### Estimate and verify the maximun n that can be solved in 1 second



In [ ]:
# Estimated time by instruction

141/200001000010

7.049964749823752e-10

In [ ]:
# Estimated number of instructions by second

1/(141/200001000010)

1418446808.5815604

Python

In [ ]:
%%writefile max_pairwise_product.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(i+1,n):                #((n**2)-n)/2
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
    if maxProduct < auxProduct:         #((n**2)-n)/2
      maxProduct = auxProduct           #[0,((n**2)-n)/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound

Writing max_pairwise_product.py


In [ ]:
!chmod +x max_pairwise_product.py

!python max_pairwise_product.py < input3.txt > output3.txt
!python max_pairwise_product.py < input4.txt > output4.txt
!python max_pairwise_product.py < input5.txt > output5.txt
!python max_pairwise_product.py < input6.txt > output6.txt
!python max_pairwise_product.py < input7.txt > output7.txt
!python max_pairwise_product.py < input8.txt > output8.txt
!python max_pairwise_product.py < input9.txt > output9.txt
!python max_pairwise_product.py < input10.txt > output10.txt

In [ ]:
import time

start_time = time.time()
!python max_pairwise_product.py < inputPy.txt

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time} segs")
print(f"Execution time: {execution_time/60} min")

399980000
Execution time: 67.44228339195251 segs
Execution time: 1.1240380565325419 min


In [ ]:
%%writefile max_pairwise_product_count.py
op = 0
xd =0
n = int(input())                        #1
op+=1
list = input().split()                  #1
op+=1
list = [int(x) for x in list]           #n
op += n

maxProduct = 0                          #1
op+=1
auxProduct = 0                          #1
op+=1
for i in range(0,n):                    #1+2n
  op+=1
  op+=2
  for j in range(i+1,n):                #((n**2)-n)/2
    op+=1
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
      op+=1
      op+=1
    op+=1
    if maxProduct < auxProduct:         #n * (n - 1) / 2
      maxProduct = auxProduct           #n * (n - 1) / 3
      op+=1

                                        # --------------For worst case
print("Result: "+str(maxProduct))
print("Number of instructions: "+str(op))                       #5 + 4n +2( n * (n - 1)) + (n * (n - 1) / 3)

Writing max_pairwise_product_count.py


In [ ]:
!python max_pairwise_product_count.py < input1.txt

Result: 90
Number of instructions: 254


In [ ]:
!python max_pairwise_product_count.py < inputPy.txt

Result: 399980000
Number of instructions: 800261148


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt

90
CPU times: user 18.9 ms, sys: 4.62 ms, total: 23.5 ms
Wall time: 923 ms


In [ ]:
%%time
!python max_pairwise_product_count.py < input1.txt

Result: 90
Number of instructions: 254
CPU times: user 5.47 ms, sys: 1.74 ms, total: 7.21 ms
Wall time: 109 ms


In [ ]:
time_per_ins = 0.0189 / 231
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.0055 / 254
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 8.181818181818182e-05
Instructions per second C++: 12222.222222222223
Time per instruction Python: 2.1653543307086614e-05
Instructions per second Python: 46181.818181818184


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [ ]:
%%writefile max_pairwise_sort_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int n;                                     // 1
    cin >> n;                                  // 1
    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              //1 + 2n
        cin >> numbers[i];                     // 1
    }

    sort(numbers.begin(), numbers.end());      // n*long
    cout << numbers[numbers.size() - 1] * numbers[numbers.size() - 2] << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_nlogn.cpp


In [ ]:
%%writefile max_pairwise_sort_nlogn.py
# Read input values
n = int(input())              # 1
numbers = list(map(int, input().split()))  # n

# Sort the list
numbers.sort()                # n*log(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 1

# Print the result
print(f"{result}")

Overwriting max_pairwise_sort_nlogn.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < input1.txt
!./max_pairwise_sort_nlogn < input2.txt

90
90


In [ ]:
!python max_pairwise_sort_nlogn.py < input1.txt
!python max_pairwise_sort_nlogn.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < input3.txt

1345294336


20 stress tests

In [ ]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

In [ ]:
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < stress_input_1.txt
!./max_pairwise_sort_nlogn < stress_input_2.txt
!./max_pairwise_sort_nlogn < stress_input_3.txt
!./max_pairwise_sort_nlogn < stress_input_4.txt
!./max_pairwise_sort_nlogn < stress_input_5.txt
!./max_pairwise_sort_nlogn < stress_input_6.txt
!./max_pairwise_sort_nlogn < stress_input_7.txt
!./max_pairwise_sort_nlogn < stress_input_8.txt
!./max_pairwise_sort_nlogn < stress_input_9.txt
!./max_pairwise_sort_nlogn < stress_input_10.txt
!./max_pairwise_sort_nlogn < stress_input_11.txt
!./max_pairwise_sort_nlogn < stress_input_12.txt
!./max_pairwise_sort_nlogn < stress_input_13.txt
!./max_pairwise_sort_nlogn < stress_input_14.txt
!./max_pairwise_sort_nlogn < stress_input_15.txt
!./max_pairwise_sort_nlogn < stress_input_16.txt
!./max_pairwise_sort_nlogn < stress_input_17.txt
!./max_pairwise_sort_nlogn < stress_input_18.txt
!./max_pairwise_sort_nlogn < stress_input_19.txt
!./max_pairwise_sort_nlogn < stress_input_20.txt

9200
9800
9108
7488
6525
7176
9024
8928
9400
8836
10000
6370
9506
8464
1504
8648
9900
7125
6417
9506


In [ ]:
!python max_pairwise_sort_nlogn.py < stress_input_1.txt
!python max_pairwise_sort_nlogn.py < stress_input_2.txt
!python max_pairwise_sort_nlogn.py < stress_input_3.txt
!python max_pairwise_sort_nlogn.py < stress_input_4.txt
!python max_pairwise_sort_nlogn.py < stress_input_5.txt
!python max_pairwise_sort_nlogn.py < stress_input_6.txt
!python max_pairwise_sort_nlogn.py < stress_input_7.txt
!python max_pairwise_sort_nlogn.py < stress_input_8.txt
!python max_pairwise_sort_nlogn.py < stress_input_9.txt
!python max_pairwise_sort_nlogn.py < stress_input_10.txt
!python max_pairwise_sort_nlogn.py < stress_input_11.txt
!python max_pairwise_sort_nlogn.py < stress_input_12.txt
!python max_pairwise_sort_nlogn.py < stress_input_13.txt
!python max_pairwise_sort_nlogn.py < stress_input_14.txt
!python max_pairwise_sort_nlogn.py < stress_input_15.txt
!python max_pairwise_sort_nlogn.py < stress_input_16.txt
!python max_pairwise_sort_nlogn.py < stress_input_17.txt
!python max_pairwise_sort_nlogn.py < stress_input_18.txt
!python max_pairwise_sort_nlogn.py < stress_input_19.txt
!python max_pairwise_sort_nlogn.py < stress_input_20.txt

9200
9800
9108
7488
6525
7176
9024
8928
9400
8836
10000
6370
9506
8464
1504
8648
9900
7125
6417
9506


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_sort_nlogn_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <bits/stdc++.h>

using namespace std;

int main() {

    uint64_t execints = 0;

    int n;                                     // 1
    execints++;
    cin >> n;                                  // 1
    execints++;
    vector<int> numbers(n);                    // n
    execints += n;

    for (int i = 0; i < n; ++i) {              //1 + 2n
        execints += 2;
        cin >> numbers[i];                     // 1
        execints++;
    }

    sort(numbers.begin(), numbers.end());      // n*long

    execints += n * log2(n);

    cout <<  numbers[n - 1] * numbers[n - 2] << "\n"; // 3

    execints += 3;

    cout << "Executed instructions: " << execints << "\n";

    return 0;                                  // 1
}

Overwriting max_pairwise_sort_nlogn_count.cpp


In [ ]:
%%writefile max_pairwise_sort_nlogn_count.py

import math


# Read input values

execints = 0
n = int(input())              # 1

execints += 1
numbers = list(map(int, input().split()))  # 4n (n input read + map o(n) + create list + cast to int values)

execints += 4 * n

# Sort the list
numbers.sort()                # n*log(n)

execints += math.log2(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 3

execints += 3

# Print the result
print(f"Executed instructions: {int(execints)}")

Overwriting max_pairwise_sort_nlogn_count.py


In [ ]:
!g++ max_pairwise_sort_nlogn_count.cpp -o max_pairwise_sort_nlogn_count
!./max_pairwise_sort_nlogn_count < input1.txt
!./max_pairwise_sort_nlogn_count < input2.txt
!./max_pairwise_sort_nlogn_count < input3.txt

90
Executed instructions: 78
90
Executed instructions: 78
1345294336
Executed instructions: 44


In [ ]:
!python max_pairwise_sort_nlogn_count.py < input1.txt
!python max_pairwise_sort_nlogn_count.py < input2.txt
!python max_pairwise_sort_nlogn_count.py < input3.txt

Executed instructions: 47
Executed instructions: 47
Executed instructions: 30


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_sort_nlogn_count.cpp -o max_pairwise_sort_nlogn_count
!./max_pairwise_sort_nlogn_count < input1.txt

90
Executed instructions: 78
CPU times: user 33.5 ms, sys: 4.76 ms, total: 38.3 ms
Wall time: 2.42 s


In [ ]:
%%time
!python max_pairwise_sort_nlogn_count.py < input1.txt

Executed instructions: 47
CPU times: user 5.34 ms, sys: 1.07 ms, total: 6.4 ms
Wall time: 106 ms


In [ ]:
time_per_ins = 0.0335 / 78
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00534 / 47
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.0004294871794871795
Instructions per second C++: 2328.3582089552237
Time per instruction Python: 0.00011361702127659575
Instructions per second Python: 8801.498127340823


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_sort_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1

    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              // 1 + 2n
        cin >> numbers[i];                     // n
    }


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = numbers[i];                           // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_efficient.cpp


In [ ]:
%%writefile max_pairwise_sort_efficient.py
print("--- Python ---")
n = int(input())
numbers = list(map(int, input().split()))

max_number = float("-inf")                   # 1
second_max_number = float("-inf")            # 1

for num in numbers:                          # 1 + 2n
    if num > max_number:                     # n
        second_max_number = max_number       # [0, n] in the worst case it would be n times
        max_number = num                     # [0, n] in the worst case it would be n times
    elif num > second_max_number:            # n
        second_max_number = num              # [0, n] in the worst case it would be n times

product = max_number * second_max_number     # 1
print(f"Product: {product}")                 # 1

Overwriting max_pairwise_sort_efficient.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < input1.txt
!./max_pairwise_sort_efficient < input2.txt

--- C++ ---
Product: 90
--- C++ ---
Product: 90


In [ ]:
! python max_pairwise_sort_efficient.py < input1.txt
! python max_pairwise_sort_efficient.py < input2.txt

--- Python ---
Product: 90
--- Python ---
Product: 90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < input3.txt
! python max_pairwise_sort_efficient.py < input3.txt

--- C++ ---
Product: 1345294336
--- Python ---
Product: 40000000000


20 stress tests

In [ ]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

In [ ]:
!g++ max_pairwise_sort_efficient.cpp -o max_pairwise_sort_efficient
!./max_pairwise_sort_efficient < stress_input_1.txt
!./max_pairwise_sort_efficient < stress_input_2.txt
!./max_pairwise_sort_efficient < stress_input_3.txt
!./max_pairwise_sort_efficient < stress_input_4.txt
!./max_pairwise_sort_efficient < stress_input_5.txt
!./max_pairwise_sort_efficient < stress_input_6.txt
!./max_pairwise_sort_efficient < stress_input_7.txt
!./max_pairwise_sort_efficient < stress_input_8.txt
!./max_pairwise_sort_efficient < stress_input_9.txt
!./max_pairwise_sort_efficient < stress_input_10.txt
!./max_pairwise_sort_efficient < stress_input_11.txt
!./max_pairwise_sort_efficient < stress_input_12.txt
!./max_pairwise_sort_efficient < stress_input_13.txt
!./max_pairwise_sort_efficient < stress_input_14.txt
!./max_pairwise_sort_efficient < stress_input_15.txt
!./max_pairwise_sort_efficient < stress_input_16.txt
!./max_pairwise_sort_efficient < stress_input_17.txt
!./max_pairwise_sort_efficient < stress_input_18.txt
!./max_pairwise_sort_efficient < stress_input_19.txt
!./max_pairwise_sort_efficient < stress_input_20.txt

--- C++ ---
Product: 7569
--- C++ ---
Product: 9702
--- C++ ---
Product: 8352
--- C++ ---
Product: 9312
--- C++ ---
Product: 9212
--- C++ ---
Product: 5832
--- C++ ---
Product: 3392
--- C++ ---
Product: 7920
--- C++ ---
Product: 8280
--- C++ ---
Product: 9215
--- C++ ---
Product: 9800
--- C++ ---
Product: 8536
--- C++ ---
Product: 9900
--- C++ ---
Product: 9200
--- C++ ---
Product: 9702
--- C++ ---
Product: 8556
--- C++ ---
Product: 5694
--- C++ ---
Product: 6400
--- C++ ---
Product: 8827
--- C++ ---
Product: 6351


In [ ]:
! python max_pairwise_sort_efficient.py < stress_input_1.txt
! python max_pairwise_sort_efficient.py < stress_input_2.txt
! python max_pairwise_sort_efficient.py < stress_input_3.txt
! python max_pairwise_sort_efficient.py < stress_input_4.txt
! python max_pairwise_sort_efficient.py < stress_input_5.txt
! python max_pairwise_sort_efficient.py < stress_input_6.txt
! python max_pairwise_sort_efficient.py < stress_input_7.txt
! python max_pairwise_sort_efficient.py < stress_input_8.txt
! python max_pairwise_sort_efficient.py < stress_input_9.txt
! python max_pairwise_sort_efficient.py < stress_input_10.txt
! python max_pairwise_sort_efficient.py < stress_input_11.txt
! python max_pairwise_sort_efficient.py < stress_input_12.txt
! python max_pairwise_sort_efficient.py < stress_input_13.txt
! python max_pairwise_sort_efficient.py < stress_input_14.txt
! python max_pairwise_sort_efficient.py < stress_input_15.txt
! python max_pairwise_sort_efficient.py < stress_input_16.txt
! python max_pairwise_sort_efficient.py < stress_input_17.txt
! python max_pairwise_sort_efficient.py < stress_input_18.txt
! python max_pairwise_sort_efficient.py < stress_input_19.txt
! python max_pairwise_sort_efficient.py < stress_input_20.txt

--- Python ---
Product: 7569
--- Python ---
Product: 9702
--- Python ---
Product: 8352
--- Python ---
Product: 9312
--- Python ---
Product: 9212
--- Python ---
Product: 5832
--- Python ---
Product: 3392
--- Python ---
Product: 7920
--- Python ---
Product: 8280
--- Python ---
Product: 9215
--- Python ---
Product: 9800
--- Python ---
Product: 8536
--- Python ---
Product: 9900
--- Python ---
Product: 9200
--- Python ---
Product: 9702
--- Python ---
Product: 8556
--- Python ---
Product: 5694
--- Python ---
Product: 6400
--- Python ---
Product: 8827
--- Python ---
Product: 6351


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_sort_efficient_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int exexins = 0;
    int n;                                     // 1
    exexins++;
    cin >> n;                                  // 1
    exexins++;
    int maxNumber = INT_MIN;                   // 1
    exexins++;
    int secondMaxNumber = INT_MIN;             // 1
    exexins++;

    vector<int> numbers(n);                    // n

    exexins += n;

    for (int i = 0; i < n; ++i) {              // 1 + 2n
        cin >> numbers[i];                     // n
        exexins++;
    }


    exexins++;
    for (int i = 0; i < n; ++i) {              //1 + 2n
        exexins += 2;

        int aux = numbers[i];                           // n
        exexins++;


        exexins++; // if
        exexins++; // else if
        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          exexins++;
          maxNumber = aux;                     // [0, n] in the worst case it would be n times
          exexins++;
        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
          exexins++;
        }

    }

    cout <<  "Executed instructions: " << exexins << " n: "<< n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_efficient_count.cpp


In [ ]:
!g++ max_pairwise_sort_efficient_count.cpp -o max_pairwise_sort_efficient_count
!./max_pairwise_sort_efficient_count < input1.txt
!./max_pairwise_sort_efficient_count < input2.txt
!./max_pairwise_sort_efficient_count < input3.txt

--- C++ ---
Executed instructions: 95 n: 10 Product: 90
--- C++ ---
Executed instructions: 78 n: 10 Product: 90
--- C++ ---
Executed instructions: 55 n: 6 Product: 1345294336


In [ ]:
%%writefile max_pairwise_sort_efficient_count.py
print("--- Python ---")
exexins = 0
n = int(input())                             # 1
exexins += 1
numbers = list(map(int, input().split()))    # 4n (n input read + map o(n) + create list + cast to int values)
exexins = 4 * n

max_number = float("-inf")                   # 1
exexins += 1
second_max_number = float("-inf")            # 1
exexins += 1

exexins += 1
for num in numbers:                          # 1 + 2n
    exexins += 2

    exexins += 1 # if
    exexins += 1 # elif
    if num > max_number:                     # n
        second_max_number = max_number       # [0, n] in the worst case it would be n times
        exexins += 1
        max_number = num                     # [0, n] in the worst case it would be n times
        exexins += 1
    elif num > second_max_number:            # n
        second_max_number = num              # [0, n] in the worst case it would be n times
        exexins += 1

product = max_number * second_max_number     # 1
exexins += 1
print(f"Executed instructions: {exexins} n: {n} Product: {product}")                 # 1

Writing max_pairwise_sort_efficient_count.py


In [ ]:
!python max_pairwise_sort_efficient_count.py < input1.txt
!python max_pairwise_sort_efficient_count.py < input2.txt
!python max_pairwise_sort_efficient_count.py < input3.txt

--- Python ---
Executed instructions: 104 n: 10 Product: 90
--- Python ---
Executed instructions: 87 n: 10 Product: 90
--- Python ---
Executed instructions: 60 n: 6 Product: 40000000000


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_sort_efficient_count.cpp -o max_pairwise_sort_efficient_count
!./max_pairwise_sort_efficient_count < input1.txt

--- C++ ---
Executed instructions: 95 n: 10 Product: 90
CPU times: user 12.3 ms, sys: 5.42 ms, total: 17.7 ms
Wall time: 912 ms


In [ ]:
%%time
!python max_pairwise_sort_efficient_count.py < input1.txt

--- Python ---
Executed instructions: 104 n: 10 Product: 90
CPU times: user 6.23 ms, sys: 5 µs, total: 6.24 ms
Wall time: 105 ms


In [ ]:
time_per_ins = 0.0123 / 85
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")
time_per_ins = 0.00623 / 104
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction C++: 0.0001447058823529412
Instructions per second C++: 6910.569105691056
Time per instruction Python: 5.9903846153846154e-05
Instructions per second Python: 16693.41894060995


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [ ]:
%%writefile max_pairwise_sort_super_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = 0;                           // n
        cin >> aux;                            // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_super_efficient.cpp


Test with inputs 1 and 2




In [ ]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < input1.txt
!./max_pairwise_sort_super_efficient < input2.txt

--- C++ ---
Product: 90
--- C++ ---
Product: 90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < input3.txt

--- C++ ---
Product: 1345294336


20 stress tests

In [ ]:
#CREATE 20 random texts
import random

for i in range(1,21):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

In [ ]:
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < stress_input_1.txt
!./max_pairwise_sort_super_efficient < stress_input_2.txt
!./max_pairwise_sort_super_efficient < stress_input_3.txt
!./max_pairwise_sort_super_efficient < stress_input_4.txt
!./max_pairwise_sort_super_efficient < stress_input_5.txt
!./max_pairwise_sort_super_efficient < stress_input_6.txt
!./max_pairwise_sort_super_efficient < stress_input_7.txt
!./max_pairwise_sort_super_efficient < stress_input_8.txt
!./max_pairwise_sort_super_efficient < stress_input_9.txt
!./max_pairwise_sort_super_efficient < stress_input_10.txt
!./max_pairwise_sort_super_efficient < stress_input_11.txt
!./max_pairwise_sort_super_efficient < stress_input_12.txt
!./max_pairwise_sort_super_efficient < stress_input_13.txt
!./max_pairwise_sort_super_efficient < stress_input_14.txt
!./max_pairwise_sort_super_efficient < stress_input_15.txt
!./max_pairwise_sort_super_efficient < stress_input_16.txt
!./max_pairwise_sort_super_efficient < stress_input_17.txt
!./max_pairwise_sort_super_efficient < stress_input_18.txt
!./max_pairwise_sort_super_efficient < stress_input_19.txt
!./max_pairwise_sort_super_efficient < stress_input_20.txt

--- C++ ---
Product: 9700
--- C++ ---
Product: 4891
--- C++ ---
Product: 7722
--- C++ ---
Product: 4422
--- C++ ---
Product: 8835
--- C++ ---
Product: 8096
--- C++ ---
Product: 8460
--- C++ ---
Product: 9600
--- C++ ---
Product: 6716
--- C++ ---
Product: 3430
--- C++ ---
Product: 9120
--- C++ ---
Product: 9600
--- C++ ---
Product: 8550
--- C++ ---
Product: 6560
--- C++ ---
Product: 9801
--- C++ ---
Product: 9120
--- C++ ---
Product: 9603
--- C++ ---
Product: 5628
--- C++ ---
Product: 8827
--- C++ ---
Product: 9405


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_sort_super_efficient_count.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    cout << "--- C++ ---" << "\n";
    int exexins = 0;
    int n;                                     // 1
    exexins++;
    cin >> n;                                  // 1
    exexins++;
    int maxNumber = INT_MIN;                   // 1
    exexins++;
    int secondMaxNumber = INT_MIN;             // 1
    exexins++;


    exexins++;
    for (int i = 0; i < n; ++i) {              //1 + 2n
        exexins += 2;
        int aux = 0;                           // n
        exexins++;

        cin>> aux;                            // n
        exexins++;


        exexins++; // if
        exexins++; // else if
        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          exexins++;
          maxNumber = aux;                     // [0, n] in the worst case it would be n times
          exexins++;
        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
          exexins++;
        }

    }

    cout <<  "Executed instructions: " << exexins << " n: "<< n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_super_efficient_count.cpp


In [ ]:
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt
!./max_pairwise_sort_super_efficient_count < input2.txt
!./max_pairwise_sort_super_efficient_count < input3.txt

--- C++ ---
Executed instructions: 85 n: 10 Product: 90
--- C++ ---
Executed instructions: 68 n: 10 Product: 90
--- C++ ---
Executed instructions: 49 n: 6 Product: 1345294336


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!g++ max_pairwise_sort_super_efficient_count.cpp -o max_pairwise_sort_super_efficient_count
!./max_pairwise_sort_super_efficient_count < input1.txt

--- C++ ---
Executed instructions: 85 n: 10 Product: 90
CPU times: user 15.6 ms, sys: 2.08 ms, total: 17.7 ms
Wall time: 813 ms


In [ ]:
time_per_ins = 0.0156 / 85
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")

Time per instruction C++: 0.00018352941176470587
Instructions per second C++: 5448.717948717949


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_sort_super_efficient.py
print("--- Python ---")
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"Product: {max_number * second_max_number}")

Overwriting max_pairwise_sort_super_efficient.py


Test with inputs 1 and 2




In [ ]:
!python max_pairwise_sort_super_efficient.py < input1.txt
!python max_pairwise_sort_super_efficient.py < input2.txt

--- Python ---
Product: 90
--- Python ---
Product: 90


Test and time it with input 3

In [ ]:
!python max_pairwise_sort_super_efficient.py < input3.txt

--- Python ---
Product: 40000000000


20 stress tests

In [ ]:
!python max_pairwise_sort_super_efficient.py < stress_input_1.txt
!python max_pairwise_sort_super_efficient.py < stress_input_2.txt
!python max_pairwise_sort_super_efficient.py < stress_input_3.txt
!python max_pairwise_sort_super_efficient.py < stress_input_4.txt
!python max_pairwise_sort_super_efficient.py < stress_input_5.txt
!python max_pairwise_sort_super_efficient.py < stress_input_6.txt
!python max_pairwise_sort_super_efficient.py < stress_input_7.txt
!python max_pairwise_sort_super_efficient.py < stress_input_8.txt
!python max_pairwise_sort_super_efficient.py < stress_input_9.txt
!python max_pairwise_sort_super_efficient.py < stress_input_10.txt
!python max_pairwise_sort_super_efficient.py < stress_input_11.txt
!python max_pairwise_sort_super_efficient.py < stress_input_12.txt
!python max_pairwise_sort_super_efficient.py < stress_input_13.txt
!python max_pairwise_sort_super_efficient.py < stress_input_14.txt
!python max_pairwise_sort_super_efficient.py < stress_input_15.txt
!python max_pairwise_sort_super_efficient.py < stress_input_16.txt
!python max_pairwise_sort_super_efficient.py < stress_input_17.txt
!python max_pairwise_sort_super_efficient.py < stress_input_18.txt
!python max_pairwise_sort_super_efficient.py < stress_input_19.txt
!python max_pairwise_sort_super_efficient.py < stress_input_20.txt

--- Python ---
Product: 9700
--- Python ---
Product: 4891
--- Python ---
Product: 7722
--- Python ---
Product: 4422
--- Python ---
Product: 8835
--- Python ---
Product: 8096
--- Python ---
Product: 8460
--- Python ---
Product: 9600
--- Python ---
Product: 6716
--- Python ---
Product: 3430
--- Python ---
Product: 9120
--- Python ---
Product: 9600
--- Python ---
Product: 8550
--- Python ---
Product: 6560
--- Python ---
Product: 9801
--- Python ---
Product: 9120
--- Python ---
Product: 9603
--- Python ---
Product: 5628
--- Python ---
Product: 8827
--- Python ---
Product: 9405


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_sort_super_efficient_count.py
print("--- Python ---")
exexins = 0
n = int(input())                            # 1
exexins += 1
max_number = float("-inf")                  # 1
exexins += 1
second_max_number = float("-inf")           # 1
exexins += 1

exexins += 3
for num in input().split():                 # 3 + 2n
    exexins += 2
    num = int(num)                          # n
    exexins += 1

    exexins += 1 # if
    exexins += 1 # elif
    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        exexins += 1
        max_number = num                    # [0, n] in the worst case it would be n times
        exexins += 1
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times
        exexins += 1

print(f"Executed instructions: {exexins} n: {n} Product: {max_number * second_max_number}")

Overwriting max_pairwise_sort_super_efficient_count.py


In [ ]:
!python max_pairwise_sort_super_efficient_count.py < input1.txt
!python max_pairwise_sort_super_efficient_count.py < input2.txt
!python max_pairwise_sort_super_efficient_count.py < input3.txt

--- Python ---
Executed instructions: 76 n: 10 Product: 90
--- Python ---
Executed instructions: 59 n: 10 Product: 90
--- Python ---
Executed instructions: 44 n: 6 Product: 40000000000


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%time
!python max_pairwise_sort_super_efficient_count.py < input1.txt

--- Python ---
Executed instructions: 76 n: 10 Product: 90
CPU times: user 7.97 ms, sys: 994 µs, total: 8.97 ms
Wall time: 107 ms


In [ ]:
time_per_ins = 0.00797 / 76
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")

Time per instruction Python: 0.00010486842105263158
Instructions per second Python: 9535.759096612297


### Estimate and verify the maximun n that can be solved in 1 second

In [ ]:
##Falta

## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [ ]:
#CREATE 100 random texts
import random

for i in range(1,101):
    fileName = f"stress_input_{i}.txt"

    with open(fileName, "w") as file:
      n = random.randint(3, 40)
      file.write(f"{n}"+ "\n")
      for i in range(1,n):
          file.write(str(random.randint(1, 100)) + " ")

#C++
##Super efficient solution (𝑇(𝑛)<10𝑛=𝑂(𝑛))

In [ ]:
%%writefile max_pairwise_sort_super_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits.h>

using namespace std;

int main() {
    int n;                                     // 1
    cin >> n;                                  // 1
    int maxNumber = INT_MIN;                   // 1
    int secondMaxNumber = INT_MIN;             // 1


    for (int i = 0; i < n; ++i) {              //1 + 2n
        int aux = 0;                           // n
        cin >> aux;                            // n

        if (aux > maxNumber){                  // n
          secondMaxNumber = maxNumber;         // [0, n] in the worst case it would be n times
          maxNumber = aux;                     // [0, n] in the worst case it would be n times

        } else if (aux > secondMaxNumber) {    // n
          secondMaxNumber = aux;               // [0, n] in the worst case it would be n times
        }

    }

    cout <<  "n: " << n << " Product: " << maxNumber * secondMaxNumber << "\n"; // 1
    return 0;                                  // 1
}

Overwriting max_pairwise_sort_super_efficient.cpp


In [ ]:
%%time
!g++ max_pairwise_sort_super_efficient.cpp -o max_pairwise_sort_super_efficient
!./max_pairwise_sort_super_efficient < stress_input_1.txt
!./max_pairwise_sort_super_efficient < stress_input_2.txt
!./max_pairwise_sort_super_efficient < stress_input_3.txt
!./max_pairwise_sort_super_efficient < stress_input_4.txt
!./max_pairwise_sort_super_efficient < stress_input_5.txt
!./max_pairwise_sort_super_efficient < stress_input_6.txt
!./max_pairwise_sort_super_efficient < stress_input_7.txt
!./max_pairwise_sort_super_efficient < stress_input_8.txt
!./max_pairwise_sort_super_efficient < stress_input_9.txt
!./max_pairwise_sort_super_efficient < stress_input_10.txt

!./max_pairwise_sort_super_efficient < stress_input_11.txt
!./max_pairwise_sort_super_efficient < stress_input_12.txt
!./max_pairwise_sort_super_efficient < stress_input_13.txt
!./max_pairwise_sort_super_efficient < stress_input_14.txt
!./max_pairwise_sort_super_efficient < stress_input_15.txt
!./max_pairwise_sort_super_efficient < stress_input_16.txt
!./max_pairwise_sort_super_efficient < stress_input_17.txt
!./max_pairwise_sort_super_efficient < stress_input_18.txt
!./max_pairwise_sort_super_efficient < stress_input_19.txt
!./max_pairwise_sort_super_efficient < stress_input_20.txt

!./max_pairwise_sort_super_efficient < stress_input_21.txt
!./max_pairwise_sort_super_efficient < stress_input_22.txt
!./max_pairwise_sort_super_efficient < stress_input_23.txt
!./max_pairwise_sort_super_efficient < stress_input_24.txt
!./max_pairwise_sort_super_efficient < stress_input_25.txt
!./max_pairwise_sort_super_efficient < stress_input_26.txt
!./max_pairwise_sort_super_efficient < stress_input_27.txt
!./max_pairwise_sort_super_efficient < stress_input_28.txt
!./max_pairwise_sort_super_efficient < stress_input_29.txt
!./max_pairwise_sort_super_efficient < stress_input_30.txt

!./max_pairwise_sort_super_efficient < stress_input_31.txt
!./max_pairwise_sort_super_efficient < stress_input_32.txt
!./max_pairwise_sort_super_efficient < stress_input_33.txt
!./max_pairwise_sort_super_efficient < stress_input_34.txt
!./max_pairwise_sort_super_efficient < stress_input_35.txt
!./max_pairwise_sort_super_efficient < stress_input_36.txt
!./max_pairwise_sort_super_efficient < stress_input_37.txt
!./max_pairwise_sort_super_efficient < stress_input_38.txt
!./max_pairwise_sort_super_efficient < stress_input_39.txt
!./max_pairwise_sort_super_efficient < stress_input_40.txt

!./max_pairwise_sort_super_efficient < stress_input_41.txt
!./max_pairwise_sort_super_efficient < stress_input_42.txt
!./max_pairwise_sort_super_efficient < stress_input_43.txt
!./max_pairwise_sort_super_efficient < stress_input_44.txt
!./max_pairwise_sort_super_efficient < stress_input_45.txt
!./max_pairwise_sort_super_efficient < stress_input_46.txt
!./max_pairwise_sort_super_efficient < stress_input_47.txt
!./max_pairwise_sort_super_efficient < stress_input_48.txt
!./max_pairwise_sort_super_efficient < stress_input_49.txt
!./max_pairwise_sort_super_efficient < stress_input_50.txt

!./max_pairwise_sort_super_efficient < stress_input_51.txt
!./max_pairwise_sort_super_efficient < stress_input_52.txt
!./max_pairwise_sort_super_efficient < stress_input_53.txt
!./max_pairwise_sort_super_efficient < stress_input_54.txt
!./max_pairwise_sort_super_efficient < stress_input_55.txt
!./max_pairwise_sort_super_efficient < stress_input_56.txt
!./max_pairwise_sort_super_efficient < stress_input_57.txt
!./max_pairwise_sort_super_efficient < stress_input_58.txt
!./max_pairwise_sort_super_efficient < stress_input_59.txt
!./max_pairwise_sort_super_efficient < stress_input_60.txt

!./max_pairwise_sort_super_efficient < stress_input_61.txt
!./max_pairwise_sort_super_efficient < stress_input_62.txt
!./max_pairwise_sort_super_efficient < stress_input_63.txt
!./max_pairwise_sort_super_efficient < stress_input_64.txt
!./max_pairwise_sort_super_efficient < stress_input_65.txt
!./max_pairwise_sort_super_efficient < stress_input_66.txt
!./max_pairwise_sort_super_efficient < stress_input_67.txt
!./max_pairwise_sort_super_efficient < stress_input_68.txt
!./max_pairwise_sort_super_efficient < stress_input_69.txt
!./max_pairwise_sort_super_efficient < stress_input_70.txt

!./max_pairwise_sort_super_efficient < stress_input_71.txt
!./max_pairwise_sort_super_efficient < stress_input_72.txt
!./max_pairwise_sort_super_efficient < stress_input_73.txt
!./max_pairwise_sort_super_efficient < stress_input_74.txt
!./max_pairwise_sort_super_efficient < stress_input_75.txt
!./max_pairwise_sort_super_efficient < stress_input_76.txt
!./max_pairwise_sort_super_efficient < stress_input_77.txt
!./max_pairwise_sort_super_efficient < stress_input_78.txt
!./max_pairwise_sort_super_efficient < stress_input_79.txt
!./max_pairwise_sort_super_efficient < stress_input_80.txt

!./max_pairwise_sort_super_efficient < stress_input_81.txt
!./max_pairwise_sort_super_efficient < stress_input_82.txt
!./max_pairwise_sort_super_efficient < stress_input_83.txt
!./max_pairwise_sort_super_efficient < stress_input_84.txt
!./max_pairwise_sort_super_efficient < stress_input_85.txt
!./max_pairwise_sort_super_efficient < stress_input_86.txt
!./max_pairwise_sort_super_efficient < stress_input_87.txt
!./max_pairwise_sort_super_efficient < stress_input_88.txt
!./max_pairwise_sort_super_efficient < stress_input_89.txt
!./max_pairwise_sort_super_efficient < stress_input_90.txt

!./max_pairwise_sort_super_efficient < stress_input_91.txt
!./max_pairwise_sort_super_efficient < stress_input_92.txt
!./max_pairwise_sort_super_efficient < stress_input_93.txt
!./max_pairwise_sort_super_efficient < stress_input_94.txt
!./max_pairwise_sort_super_efficient < stress_input_95.txt
!./max_pairwise_sort_super_efficient < stress_input_96.txt
!./max_pairwise_sort_super_efficient < stress_input_97.txt
!./max_pairwise_sort_super_efficient < stress_input_98.txt
!./max_pairwise_sort_super_efficient < stress_input_99.txt
!./max_pairwise_sort_super_efficient < stress_input_100.txt

n: 12 Product: 4118
n: 39 Product: 9408
n: 21 Product: 10000
n: 29 Product: 8463
n: 7 Product: 3750
n: 25 Product: 9800
n: 16 Product: 9506
n: 36 Product: 8277
n: 8 Product: 6716
n: 19 Product: 9200
n: 4 Product: 1827
n: 13 Product: 7546
n: 8 Product: 3363
n: 23 Product: 9900
n: 33 Product: 9024
n: 30 Product: 9310
n: 5 Product: 8700
n: 34 Product: 9408
n: 15 Product: 9215
n: 6 Product: 5626
n: 29 Product: 9700
n: 32 Product: 9016
n: 20 Product: 9506
n: 38 Product: 9800
n: 39 Product: 9506
n: 30 Product: 8928
n: 10 Product: 8256
n: 40 Product: 8372
n: 26 Product: 9200
n: 3 Product: 800
n: 5 Product: 3325
n: 14 Product: 8544
n: 22 Product: 9300
n: 15 Product: 8930
n: 4 Product: 6603
n: 40 Product: 9700
n: 22 Product: 9702
n: 34 Product: 9409
n: 31 Product: 8910
n: 37 Product: 9900
n: 37 Product: 9400
n: 27 Product: 9504
n: 32 Product: 9312
n: 13 Product: 8360
n: 16 Product: 6300
n: 5 Product: 6786
n: 34 Product: 7482
n: 20 Product: 9900
n: 6 Product: 2294
n: 19 Product: 9600
n: 35 Produ

##Sortin solution (𝑇(𝑛)=𝑂(𝑛𝑙𝑛𝑔))

In [ ]:
%%writefile max_pairwise_sort_nlogn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    int n;                                     // 1
    cin >> n;                                  // 1
    vector<int> numbers(n);                    // n

    for (int i = 0; i < n; ++i) {              //1 + 2n
        cin >> numbers[i];                     // 1
    }

    sort(numbers.begin(), numbers.end());      // n*long
    cout <<  "n: " << n << " Product: " << numbers[numbers.size() - 1] * numbers[numbers.size() - 2] << "\n"; // 1
    return 0;                                  // 1
}

Writing max_pairwise_sort_nlogn.cpp


In [ ]:
%%time
!g++ max_pairwise_sort_nlogn.cpp -o max_pairwise_sort_nlogn
!./max_pairwise_sort_nlogn < stress_input_1.txt
!./max_pairwise_sort_nlogn < stress_input_2.txt
!./max_pairwise_sort_nlogn < stress_input_3.txt
!./max_pairwise_sort_nlogn < stress_input_4.txt
!./max_pairwise_sort_nlogn < stress_input_5.txt
!./max_pairwise_sort_nlogn < stress_input_6.txt
!./max_pairwise_sort_nlogn < stress_input_7.txt
!./max_pairwise_sort_nlogn < stress_input_8.txt
!./max_pairwise_sort_nlogn < stress_input_9.txt
!./max_pairwise_sort_nlogn < stress_input_10.txt

!./max_pairwise_sort_nlogn < stress_input_11.txt
!./max_pairwise_sort_nlogn < stress_input_12.txt
!./max_pairwise_sort_nlogn < stress_input_13.txt
!./max_pairwise_sort_nlogn < stress_input_14.txt
!./max_pairwise_sort_nlogn < stress_input_15.txt
!./max_pairwise_sort_nlogn < stress_input_16.txt
!./max_pairwise_sort_nlogn < stress_input_17.txt
!./max_pairwise_sort_nlogn < stress_input_18.txt
!./max_pairwise_sort_nlogn < stress_input_19.txt
!./max_pairwise_sort_nlogn < stress_input_20.txt

!./max_pairwise_sort_nlogn < stress_input_21.txt
!./max_pairwise_sort_nlogn < stress_input_22.txt
!./max_pairwise_sort_nlogn < stress_input_23.txt
!./max_pairwise_sort_nlogn < stress_input_24.txt
!./max_pairwise_sort_nlogn < stress_input_25.txt
!./max_pairwise_sort_nlogn < stress_input_26.txt
!./max_pairwise_sort_nlogn < stress_input_27.txt
!./max_pairwise_sort_nlogn < stress_input_28.txt
!./max_pairwise_sort_nlogn < stress_input_29.txt
!./max_pairwise_sort_nlogn < stress_input_30.txt

!./max_pairwise_sort_nlogn < stress_input_31.txt
!./max_pairwise_sort_nlogn < stress_input_32.txt
!./max_pairwise_sort_nlogn < stress_input_33.txt
!./max_pairwise_sort_nlogn < stress_input_34.txt
!./max_pairwise_sort_nlogn < stress_input_35.txt
!./max_pairwise_sort_nlogn < stress_input_36.txt
!./max_pairwise_sort_nlogn < stress_input_37.txt
!./max_pairwise_sort_nlogn < stress_input_38.txt
!./max_pairwise_sort_nlogn < stress_input_39.txt
!./max_pairwise_sort_nlogn < stress_input_40.txt

!./max_pairwise_sort_nlogn < stress_input_41.txt
!./max_pairwise_sort_nlogn < stress_input_42.txt
!./max_pairwise_sort_nlogn < stress_input_43.txt
!./max_pairwise_sort_nlogn < stress_input_44.txt
!./max_pairwise_sort_nlogn < stress_input_45.txt
!./max_pairwise_sort_nlogn < stress_input_46.txt
!./max_pairwise_sort_nlogn < stress_input_47.txt
!./max_pairwise_sort_nlogn < stress_input_48.txt
!./max_pairwise_sort_nlogn < stress_input_49.txt
!./max_pairwise_sort_nlogn < stress_input_50.txt

!./max_pairwise_sort_nlogn < stress_input_51.txt
!./max_pairwise_sort_nlogn < stress_input_52.txt
!./max_pairwise_sort_nlogn < stress_input_53.txt
!./max_pairwise_sort_nlogn < stress_input_54.txt
!./max_pairwise_sort_nlogn < stress_input_55.txt
!./max_pairwise_sort_nlogn < stress_input_56.txt
!./max_pairwise_sort_nlogn < stress_input_57.txt
!./max_pairwise_sort_nlogn < stress_input_58.txt
!./max_pairwise_sort_nlogn < stress_input_59.txt
!./max_pairwise_sort_nlogn < stress_input_60.txt

!./max_pairwise_sort_nlogn < stress_input_61.txt
!./max_pairwise_sort_nlogn < stress_input_62.txt
!./max_pairwise_sort_nlogn < stress_input_63.txt
!./max_pairwise_sort_nlogn < stress_input_64.txt
!./max_pairwise_sort_nlogn < stress_input_65.txt
!./max_pairwise_sort_nlogn < stress_input_66.txt
!./max_pairwise_sort_nlogn < stress_input_67.txt
!./max_pairwise_sort_nlogn < stress_input_68.txt
!./max_pairwise_sort_nlogn < stress_input_69.txt
!./max_pairwise_sort_nlogn < stress_input_70.txt

!./max_pairwise_sort_nlogn < stress_input_71.txt
!./max_pairwise_sort_nlogn < stress_input_72.txt
!./max_pairwise_sort_nlogn < stress_input_73.txt
!./max_pairwise_sort_nlogn < stress_input_74.txt
!./max_pairwise_sort_nlogn < stress_input_75.txt
!./max_pairwise_sort_nlogn < stress_input_76.txt
!./max_pairwise_sort_nlogn < stress_input_77.txt
!./max_pairwise_sort_nlogn < stress_input_78.txt
!./max_pairwise_sort_nlogn < stress_input_79.txt
!./max_pairwise_sort_nlogn < stress_input_80.txt

!./max_pairwise_sort_nlogn < stress_input_81.txt
!./max_pairwise_sort_nlogn < stress_input_82.txt
!./max_pairwise_sort_nlogn < stress_input_83.txt
!./max_pairwise_sort_nlogn < stress_input_84.txt
!./max_pairwise_sort_nlogn < stress_input_85.txt
!./max_pairwise_sort_nlogn < stress_input_86.txt
!./max_pairwise_sort_nlogn < stress_input_87.txt
!./max_pairwise_sort_nlogn < stress_input_88.txt
!./max_pairwise_sort_nlogn < stress_input_89.txt
!./max_pairwise_sort_nlogn < stress_input_90.txt

!./max_pairwise_sort_nlogn < stress_input_91.txt
!./max_pairwise_sort_nlogn < stress_input_92.txt
!./max_pairwise_sort_nlogn < stress_input_93.txt
!./max_pairwise_sort_nlogn < stress_input_94.txt
!./max_pairwise_sort_nlogn < stress_input_95.txt
!./max_pairwise_sort_nlogn < stress_input_96.txt
!./max_pairwise_sort_nlogn < stress_input_97.txt
!./max_pairwise_sort_nlogn < stress_input_98.txt
!./max_pairwise_sort_nlogn < stress_input_99.txt
!./max_pairwise_sort_nlogn < stress_input_100.txt

n: 12 Product: 4118
n: 39 Product: 9408
n: 21 Product: 10000
n: 29 Product: 8463
n: 7 Product: 3750
n: 25 Product: 9800
n: 16 Product: 9506
n: 36 Product: 8277
n: 8 Product: 6716
n: 19 Product: 9200
n: 4 Product: 1827
n: 13 Product: 7546
n: 8 Product: 3363
n: 23 Product: 9900
n: 33 Product: 9024
n: 30 Product: 9310
n: 5 Product: 8700
n: 34 Product: 9408
n: 15 Product: 9215
n: 6 Product: 5626
n: 29 Product: 9700
n: 32 Product: 9016
n: 20 Product: 9506
n: 38 Product: 9800
n: 39 Product: 9506
n: 30 Product: 8928
n: 10 Product: 8256
n: 40 Product: 8372
n: 26 Product: 9200
n: 3 Product: 800
n: 5 Product: 3325
n: 14 Product: 8544
n: 22 Product: 9300
n: 15 Product: 8930
n: 4 Product: 6603
n: 40 Product: 9700
n: 22 Product: 9702
n: 34 Product: 9409
n: 31 Product: 8910
n: 37 Product: 9900
n: 37 Product: 9400
n: 27 Product: 9504
n: 32 Product: 9312
n: 13 Product: 8360
n: 16 Product: 6300
n: 5 Product: 6786
n: 34 Product: 7482
n: 20 Product: 9900
n: 6 Product: 2294
n: 19 Product: 9600
n: 35 Produ

#Python
##Super efficient solution (𝑇(𝑛)<10𝑛=𝑂(𝑛))

In [ ]:
%%writefile max_pairwise_sort_super_efficient.py
n = int(input())                            # 1
max_number = float("-inf")                  # 1
second_max_number = float("-inf")           # 1

for num in input().split():                 # 3 + 2n

    num = int(num)                          # n

    if num > max_number:                    # n
        second_max_number = max_number      # [0, n] in the worst case it would be n times
        max_number = num                    # [0, n] in the worst case it would be n times
    elif num > second_max_number:           # n
        second_max_number = num             # [0, n] in the worst case it would be n times

print(f"n: {n} Product: {max_number * second_max_number}")

Overwriting max_pairwise_sort_super_efficient.py


In [ ]:
%%time
!python max_pairwise_sort_super_efficient.py < stress_input_1.txt
!python max_pairwise_sort_super_efficient.py < stress_input_2.txt
!python max_pairwise_sort_super_efficient.py < stress_input_3.txt
!python max_pairwise_sort_super_efficient.py < stress_input_4.txt
!python max_pairwise_sort_super_efficient.py < stress_input_5.txt
!python max_pairwise_sort_super_efficient.py < stress_input_6.txt
!python max_pairwise_sort_super_efficient.py < stress_input_7.txt
!python max_pairwise_sort_super_efficient.py < stress_input_8.txt
!python max_pairwise_sort_super_efficient.py < stress_input_9.txt
!python max_pairwise_sort_super_efficient.py < stress_input_10.txt

!python max_pairwise_sort_super_efficient.py < stress_input_11.txt
!python max_pairwise_sort_super_efficient.py < stress_input_12.txt
!python max_pairwise_sort_super_efficient.py < stress_input_13.txt
!python max_pairwise_sort_super_efficient.py < stress_input_14.txt
!python max_pairwise_sort_super_efficient.py < stress_input_15.txt
!python max_pairwise_sort_super_efficient.py < stress_input_16.txt
!python max_pairwise_sort_super_efficient.py < stress_input_17.txt
!python max_pairwise_sort_super_efficient.py < stress_input_18.txt
!python max_pairwise_sort_super_efficient.py < stress_input_19.txt
!python max_pairwise_sort_super_efficient.py < stress_input_20.txt

!python max_pairwise_sort_super_efficient.py < stress_input_21.txt
!python max_pairwise_sort_super_efficient.py < stress_input_22.txt
!python max_pairwise_sort_super_efficient.py < stress_input_23.txt
!python max_pairwise_sort_super_efficient.py < stress_input_24.txt
!python max_pairwise_sort_super_efficient.py < stress_input_25.txt
!python max_pairwise_sort_super_efficient.py < stress_input_26.txt
!python max_pairwise_sort_super_efficient.py < stress_input_27.txt
!python max_pairwise_sort_super_efficient.py < stress_input_28.txt
!python max_pairwise_sort_super_efficient.py < stress_input_29.txt
!python max_pairwise_sort_super_efficient.py < stress_input_30.txt

!python max_pairwise_sort_super_efficient.py < stress_input_31.txt
!python max_pairwise_sort_super_efficient.py < stress_input_32.txt
!python max_pairwise_sort_super_efficient.py < stress_input_33.txt
!python max_pairwise_sort_super_efficient.py < stress_input_34.txt
!python max_pairwise_sort_super_efficient.py < stress_input_35.txt
!python max_pairwise_sort_super_efficient.py < stress_input_36.txt
!python max_pairwise_sort_super_efficient.py < stress_input_37.txt
!python max_pairwise_sort_super_efficient.py < stress_input_38.txt
!python max_pairwise_sort_super_efficient.py < stress_input_39.txt
!python max_pairwise_sort_super_efficient.py < stress_input_40.txt

!python max_pairwise_sort_super_efficient.py < stress_input_41.txt
!python max_pairwise_sort_super_efficient.py < stress_input_42.txt
!python max_pairwise_sort_super_efficient.py < stress_input_43.txt
!python max_pairwise_sort_super_efficient.py < stress_input_44.txt
!python max_pairwise_sort_super_efficient.py < stress_input_45.txt
!python max_pairwise_sort_super_efficient.py < stress_input_46.txt
!python max_pairwise_sort_super_efficient.py < stress_input_47.txt
!python max_pairwise_sort_super_efficient.py < stress_input_48.txt
!python max_pairwise_sort_super_efficient.py < stress_input_49.txt
!python max_pairwise_sort_super_efficient.py < stress_input_50.txt

!python max_pairwise_sort_super_efficient.py < stress_input_51.txt
!python max_pairwise_sort_super_efficient.py < stress_input_52.txt
!python max_pairwise_sort_super_efficient.py < stress_input_53.txt
!python max_pairwise_sort_super_efficient.py < stress_input_54.txt
!python max_pairwise_sort_super_efficient.py < stress_input_55.txt
!python max_pairwise_sort_super_efficient.py < stress_input_56.txt
!python max_pairwise_sort_super_efficient.py < stress_input_57.txt
!python max_pairwise_sort_super_efficient.py < stress_input_58.txt
!python max_pairwise_sort_super_efficient.py < stress_input_59.txt
!python max_pairwise_sort_super_efficient.py < stress_input_60.txt

!python max_pairwise_sort_super_efficient.py < stress_input_61.txt
!python max_pairwise_sort_super_efficient.py < stress_input_62.txt
!python max_pairwise_sort_super_efficient.py < stress_input_63.txt
!python max_pairwise_sort_super_efficient.py < stress_input_64.txt
!python max_pairwise_sort_super_efficient.py < stress_input_65.txt
!python max_pairwise_sort_super_efficient.py < stress_input_66.txt
!python max_pairwise_sort_super_efficient.py < stress_input_67.txt
!python max_pairwise_sort_super_efficient.py < stress_input_68.txt
!python max_pairwise_sort_super_efficient.py < stress_input_69.txt
!python max_pairwise_sort_super_efficient.py < stress_input_70.txt

!python max_pairwise_sort_super_efficient.py < stress_input_71.txt
!python max_pairwise_sort_super_efficient.py < stress_input_72.txt
!python max_pairwise_sort_super_efficient.py < stress_input_73.txt
!python max_pairwise_sort_super_efficient.py < stress_input_74.txt
!python max_pairwise_sort_super_efficient.py < stress_input_75.txt
!python max_pairwise_sort_super_efficient.py < stress_input_76.txt
!python max_pairwise_sort_super_efficient.py < stress_input_77.txt
!python max_pairwise_sort_super_efficient.py < stress_input_78.txt
!python max_pairwise_sort_super_efficient.py < stress_input_79.txt
!python max_pairwise_sort_super_efficient.py < stress_input_80.txt

!python max_pairwise_sort_super_efficient.py < stress_input_81.txt
!python max_pairwise_sort_super_efficient.py < stress_input_82.txt
!python max_pairwise_sort_super_efficient.py < stress_input_83.txt
!python max_pairwise_sort_super_efficient.py < stress_input_84.txt
!python max_pairwise_sort_super_efficient.py < stress_input_85.txt
!python max_pairwise_sort_super_efficient.py < stress_input_86.txt
!python max_pairwise_sort_super_efficient.py < stress_input_87.txt
!python max_pairwise_sort_super_efficient.py < stress_input_88.txt
!python max_pairwise_sort_super_efficient.py < stress_input_89.txt
!python max_pairwise_sort_super_efficient.py < stress_input_90.txt

!python max_pairwise_sort_super_efficient.py < stress_input_91.txt
!python max_pairwise_sort_super_efficient.py < stress_input_92.txt
!python max_pairwise_sort_super_efficient.py < stress_input_93.txt
!python max_pairwise_sort_super_efficient.py < stress_input_94.txt
!python max_pairwise_sort_super_efficient.py < stress_input_95.txt
!python max_pairwise_sort_super_efficient.py < stress_input_96.txt
!python max_pairwise_sort_super_efficient.py < stress_input_97.txt
!python max_pairwise_sort_super_efficient.py < stress_input_98.txt
!python max_pairwise_sort_super_efficient.py < stress_input_99.txt
!python max_pairwise_sort_super_efficient.py < stress_input_100.txt

n: 12 Product: 4118
n: 39 Product: 9408
n: 21 Product: 10000
n: 29 Product: 8463
n: 7 Product: 3750
n: 25 Product: 9800
n: 16 Product: 9506
n: 36 Product: 8277
n: 8 Product: 6716
n: 19 Product: 9200
n: 4 Product: 1827
n: 13 Product: 7546
n: 8 Product: 3363
n: 23 Product: 9900
n: 33 Product: 9024
n: 30 Product: 9310
n: 5 Product: 8700
n: 34 Product: 9408
n: 15 Product: 9215
n: 6 Product: 5626
n: 29 Product: 9700
n: 32 Product: 9016
n: 20 Product: 9506
n: 38 Product: 9800
n: 39 Product: 9506
n: 30 Product: 8928
n: 10 Product: 8256
n: 40 Product: 8372
n: 26 Product: 9200
n: 3 Product: 800
n: 5 Product: 3325
n: 14 Product: 8544
n: 22 Product: 9300
n: 15 Product: 8930
n: 4 Product: 6603
n: 40 Product: 9700
n: 22 Product: 9702
n: 34 Product: 9409
n: 31 Product: 8910
n: 37 Product: 9900
n: 37 Product: 9400
n: 27 Product: 9504
n: 32 Product: 9312
n: 13 Product: 8360
n: 16 Product: 6300
n: 5 Product: 6786
n: 34 Product: 7482
n: 20 Product: 9900
n: 6 Product: 2294
n: 19 Product: 9600
n: 35 Produ

##Sortin solution (𝑇(𝑛)=𝑂(𝑛𝑙𝑛𝑔))

In [ ]:
%%writefile max_pairwise_sort_nlogn.py
# Read input values
n = int(input())              # 1
numbers = list(map(int, input().split()))  # n

# Sort the list
numbers.sort()                # n*log(n)

# Calculate the maximum pairwise product
result = numbers[-1] * numbers[-2]  # 1

# Print the result
print(f"n: {n} Product: {result}")

Writing max_pairwise_sort_nlogn.py


In [ ]:
%%time
!python max_pairwise_sort_nlogn.py < stress_input_1.txt
!python max_pairwise_sort_nlogn.py < stress_input_2.txt
!python max_pairwise_sort_nlogn.py < stress_input_3.txt
!python max_pairwise_sort_nlogn.py < stress_input_4.txt
!python max_pairwise_sort_nlogn.py < stress_input_5.txt
!python max_pairwise_sort_nlogn.py < stress_input_6.txt
!python max_pairwise_sort_nlogn.py < stress_input_7.txt
!python max_pairwise_sort_nlogn.py < stress_input_8.txt
!python max_pairwise_sort_nlogn.py < stress_input_9.txt
!python max_pairwise_sort_nlogn.py < stress_input_10.txt

!python max_pairwise_sort_nlogn.py < stress_input_11.txt
!python max_pairwise_sort_nlogn.py < stress_input_12.txt
!python max_pairwise_sort_nlogn.py < stress_input_13.txt
!python max_pairwise_sort_nlogn.py < stress_input_14.txt
!python max_pairwise_sort_nlogn.py < stress_input_15.txt
!python max_pairwise_sort_nlogn.py < stress_input_16.txt
!python max_pairwise_sort_nlogn.py < stress_input_17.txt
!python max_pairwise_sort_nlogn.py < stress_input_18.txt
!python max_pairwise_sort_nlogn.py < stress_input_19.txt
!python max_pairwise_sort_nlogn.py < stress_input_20.txt

!python max_pairwise_sort_nlogn.py < stress_input_21.txt
!python max_pairwise_sort_nlogn.py < stress_input_22.txt
!python max_pairwise_sort_nlogn.py < stress_input_23.txt
!python max_pairwise_sort_nlogn.py < stress_input_24.txt
!python max_pairwise_sort_nlogn.py < stress_input_25.txt
!python max_pairwise_sort_nlogn.py < stress_input_26.txt
!python max_pairwise_sort_nlogn.py < stress_input_27.txt
!python max_pairwise_sort_nlogn.py < stress_input_28.txt
!python max_pairwise_sort_nlogn.py < stress_input_29.txt
!python max_pairwise_sort_nlogn.py < stress_input_30.txt

!python max_pairwise_sort_nlogn.py < stress_input_31.txt
!python max_pairwise_sort_nlogn.py < stress_input_32.txt
!python max_pairwise_sort_nlogn.py < stress_input_33.txt
!python max_pairwise_sort_nlogn.py < stress_input_34.txt
!python max_pairwise_sort_nlogn.py < stress_input_35.txt
!python max_pairwise_sort_nlogn.py < stress_input_36.txt
!python max_pairwise_sort_nlogn.py < stress_input_37.txt
!python max_pairwise_sort_nlogn.py < stress_input_38.txt
!python max_pairwise_sort_nlogn.py < stress_input_39.txt
!python max_pairwise_sort_nlogn.py < stress_input_40.txt

!python max_pairwise_sort_nlogn.py < stress_input_41.txt
!python max_pairwise_sort_nlogn.py < stress_input_42.txt
!python max_pairwise_sort_nlogn.py < stress_input_43.txt
!python max_pairwise_sort_nlogn.py < stress_input_44.txt
!python max_pairwise_sort_nlogn.py < stress_input_45.txt
!python max_pairwise_sort_nlogn.py < stress_input_46.txt
!python max_pairwise_sort_nlogn.py < stress_input_47.txt
!python max_pairwise_sort_nlogn.py < stress_input_48.txt
!python max_pairwise_sort_nlogn.py < stress_input_49.txt
!python max_pairwise_sort_nlogn.py < stress_input_50.txt

!python max_pairwise_sort_nlogn.py < stress_input_51.txt
!python max_pairwise_sort_nlogn.py < stress_input_52.txt
!python max_pairwise_sort_nlogn.py < stress_input_53.txt
!python max_pairwise_sort_nlogn.py < stress_input_54.txt
!python max_pairwise_sort_nlogn.py < stress_input_55.txt
!python max_pairwise_sort_nlogn.py < stress_input_56.txt
!python max_pairwise_sort_nlogn.py < stress_input_57.txt
!python max_pairwise_sort_nlogn.py < stress_input_58.txt
!python max_pairwise_sort_nlogn.py < stress_input_59.txt
!python max_pairwise_sort_nlogn.py < stress_input_60.txt

!python max_pairwise_sort_nlogn.py < stress_input_61.txt
!python max_pairwise_sort_nlogn.py < stress_input_62.txt
!python max_pairwise_sort_nlogn.py < stress_input_63.txt
!python max_pairwise_sort_nlogn.py < stress_input_64.txt
!python max_pairwise_sort_nlogn.py < stress_input_65.txt
!python max_pairwise_sort_nlogn.py < stress_input_66.txt
!python max_pairwise_sort_nlogn.py < stress_input_67.txt
!python max_pairwise_sort_nlogn.py < stress_input_68.txt
!python max_pairwise_sort_nlogn.py < stress_input_69.txt
!python max_pairwise_sort_nlogn.py < stress_input_70.txt

!python max_pairwise_sort_nlogn.py < stress_input_71.txt
!python max_pairwise_sort_nlogn.py < stress_input_72.txt
!python max_pairwise_sort_nlogn.py < stress_input_73.txt
!python max_pairwise_sort_nlogn.py < stress_input_74.txt
!python max_pairwise_sort_nlogn.py < stress_input_75.txt
!python max_pairwise_sort_nlogn.py < stress_input_76.txt
!python max_pairwise_sort_nlogn.py < stress_input_77.txt
!python max_pairwise_sort_nlogn.py < stress_input_78.txt
!python max_pairwise_sort_nlogn.py < stress_input_79.txt
!python max_pairwise_sort_nlogn.py < stress_input_80.txt

!python max_pairwise_sort_nlogn.py < stress_input_81.txt
!python max_pairwise_sort_nlogn.py < stress_input_82.txt
!python max_pairwise_sort_nlogn.py < stress_input_83.txt
!python max_pairwise_sort_nlogn.py < stress_input_84.txt
!python max_pairwise_sort_nlogn.py < stress_input_85.txt
!python max_pairwise_sort_nlogn.py < stress_input_86.txt
!python max_pairwise_sort_nlogn.py < stress_input_87.txt
!python max_pairwise_sort_nlogn.py < stress_input_88.txt
!python max_pairwise_sort_nlogn.py < stress_input_89.txt
!python max_pairwise_sort_nlogn.py < stress_input_90.txt

!python max_pairwise_sort_nlogn.py < stress_input_91.txt
!python max_pairwise_sort_nlogn.py < stress_input_92.txt
!python max_pairwise_sort_nlogn.py < stress_input_93.txt
!python max_pairwise_sort_nlogn.py < stress_input_94.txt
!python max_pairwise_sort_nlogn.py < stress_input_95.txt
!python max_pairwise_sort_nlogn.py < stress_input_96.txt
!python max_pairwise_sort_nlogn.py < stress_input_97.txt
!python max_pairwise_sort_nlogn.py < stress_input_98.txt
!python max_pairwise_sort_nlogn.py < stress_input_99.txt
!python max_pairwise_sort_nlogn.py < stress_input_100.txt

n: 12 Product: 4118
n: 39 Product: 9408
n: 21 Product: 10000
n: 29 Product: 8463
n: 7 Product: 3750
n: 25 Product: 9800
n: 16 Product: 9506
n: 36 Product: 8277
n: 8 Product: 6716
n: 19 Product: 9200
n: 4 Product: 1827
n: 13 Product: 7546
n: 8 Product: 3363
n: 23 Product: 9900
n: 33 Product: 9024
n: 30 Product: 9310
n: 5 Product: 8700
n: 34 Product: 9408
n: 15 Product: 9215
n: 6 Product: 5626
n: 29 Product: 9700
n: 32 Product: 9016
n: 20 Product: 9506
n: 38 Product: 9800
n: 39 Product: 9506
n: 30 Product: 8928
n: 10 Product: 8256
n: 40 Product: 8372
n: 26 Product: 9200
n: 3 Product: 800
n: 5 Product: 3325
n: 14 Product: 8544
n: 22 Product: 9300
n: 15 Product: 8930
n: 4 Product: 6603
n: 40 Product: 9700
n: 22 Product: 9702
n: 34 Product: 9409
n: 31 Product: 8910
n: 37 Product: 9900
n: 37 Product: 9400
n: 27 Product: 9504
n: 32 Product: 9312
n: 13 Product: 8360
n: 16 Product: 6300
n: 5 Product: 6786
n: 34 Product: 7482
n: 20 Product: 9900
n: 6 Product: 2294
n: 19 Product: 9600
n: 35 Produ

#Resultados

In [ ]:
# Import the Pandas library
import pandas as pd

# Create a sample DataFrame
data = {'Language': ['C++', 'Python'],
        'Super effienct (ms)': [573, 553],
        'Sort solution (ms)': [545, 509]}

df = pd.DataFrame(data)

# Display the table
display(df)

Language  Super effienct (ms)  Sort solution (ms)
0      C++                  573                 545
1   Python                  553                 509